common/src/backends/otlp_tracing_bridge.rs

```diff
@@ -152,3 +153,104 @@ const fn severity_of_level(level: &Level) -> Severity {
         Level::ERROR => Severity::Error,
     }
 }
+
+pub struct ErrorTracingLayer<S> {
+    _registry: std::marker::PhantomData<S>,
+}
+
+impl<S> ErrorTracingLayer<S>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+{
+    pub fn new() -> Self {
+        ErrorTracingLayer {
+            _registry: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<S> Layer<S> for ErrorTracingLayer<S>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+{
+    fn on_event(
+        &self,
+        event: &tracing_core::Event<'_>,
+        ctx: tracing_subscriber::layer::Context<'_, S>,
+    ) {
+        // We only care about error events.
+        if !ErrorVisitor::is_valid(event.metadata()) {
+            return;
+        }
+
+        let mut visitor = ErrorVisitor::default();
+        event.record(&mut visitor);
+
+        let DatadogError {
+            message,
+            r#type,
+            stack,
+        } = visitor.error;
+
+        if let Some(span) = ctx.lookup_current() {
+            if let Some(otel_data) = span.extensions_mut().get_mut::<OtelData>() {
+                let error_fields = [
+                    KeyValue::new("error.message", message),
+                    KeyValue::new("error.type", r#type),
+                    KeyValue::new("error.stack", stack),
+                ];
+                let builder_attrs = otel_data
+                    .builder
+                    .attributes
+                    .get_or_insert(Vec::with_capacity(3));
+                builder_attrs.extend(error_fields);
+            }
+        };
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct DatadogError {
+    pub message: String,
+    pub r#type: String,
+    pub stack: String,
+}
+#[derive(Default)]
+struct ErrorVisitor {
+    error: DatadogError,
+}
+
+impl ErrorVisitor {
+    /// We only care about error events with the [ID_IDENT] field.
+    fn is_valid(metadata: &Metadata) -> bool {
+        metadata.is_event() && metadata.level() == &Level::ERROR
+    }
+}
+
+impl Visit for ErrorVisitor {
+    fn record_debug(&mut self, _field: &Field, _value: &dyn std::fmt::Debug) {
+        // This visitor is only concerned with recording errors, do nothing for debug fields.
+    }
+    fn record_error(&mut self, _field: &Field, value: &(dyn std::error::Error + 'static)) {
+        // Create an error source chain, including the top-level error.
+        let source_chain = {
+            // Datadog expects there to be at least two lines in the stack field for the apm error
+            // tracking feature to work, so we ensure there always is.
+            let mut chain: String = format!("Error source chain:\n{}", value);
+            let mut next_err = value.source();
+            // TODO: skip the first error?
```

Comment:
what do you mean?

Reply:
Also why?

Reply:
My thinking was that maybe we want to skip the first error, since it will be in the error message field. But I decided against it, I thought I had removed this todo actually, so I will do it now. :smile: